pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -292,24 +292,30 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
 
         // Connection is gone, close the producers immediately
         producers.forEach((__, producerFuture) -> {
+            // prevent race conditions in completing producers
+            if (!producerFuture.isDone()
+                    && producerFuture.completeExceptionally(new IllegalStateException("Connection closed."))) {
```

Comment:
Nit. I don't think we need to call `!producerFuture.isDone()`. The logic is the same when removing the check.

```suggestion
            if (producerFuture.completeExceptionally(new IllegalStateException("Connection closed."))) {
```

Reply:
no, it's different. Creating an exception is a heavy operation and the original version prevents that from happening if it's done.

Reply:
@lhotari - maybe you were trying to avoid creating the `IllegalStateException`? Maybe my proposed change isn't better. Let me know what you think.

Reply:
Yes exactly.

Reply:
Thanks for confirming. Makes sense.